Update data inlines

phausler opened the pull request:
This is just an experiment for now to test how Data inlines from a performance standpoint.

phausler commented:
@swift-ci please performance test

phausler commented:
@swift-ci please benchmark

slavapestov commented on stdlib/public/SDK/Foundation/Data.swift:
```diff
@@ -105,20 +107,28 @@ public final class _DataStorage {
-        @inline(__always)
+    @usableFromInline
+    var bytes: UnsafeRawPointer? {
+        @inlinable
```
You can declare the property itself @inlinable instead if you want

phausler commented:
@swift-ci please benchmark

phausler commented:
@swift-ci please smoke test

phausler commented:
@swift-ci please benchmark

phausler commented:
@swift-ci please smoke test

phausler commented:
@swift-ci please benchmark

phausler commented:
@airspeedswift DataAppendSequence uses the hidden api for sequences. Do you happen to know who might have insight on why this inline change adversely affects this particular test?

airspeedswift commented:
The problem isn't the underscored method. It's that you're looping over a generic sequence at the end in uninlineable code. That is inevitably going to be massively slower, because the compiler can't specialize, in the case of this benchmark, that it's got a `Repeated` sequence that is essentially just a constant and a counter. If it can specialize, that should optimize down to just some tiny number of simple instructions. But without inlining, it needs to go through a fully generic call to `next()` over and over again.

phausler commented:
@swift-ci please benchmark

phausler commented:
oh I was just rolling it back to the original to verify that fixed the benchmark.

phausler commented:
@swift-ci please benchmark

phausler commented:
@swift-ci please smoke test

phausler commented:
@swift-ci please benchmark

phausler commented:
@swift-ci please smoke test

parkera commented on stdlib/public/SDK/Foundation/Data.swift:
```diff
@@ -1461,56 +1477,55 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
-                append(byte)
-            }
+    }
+
```
This is the part we'll have to remove due to test failures?

phausler committed:
[Foundation] Migrate Data to a ABI stable variant

This has one known performance regression; appending with a sequence (not an Array, but something like a Repeated<UInt8>) is slower. We aknowlege this regression and plan to offer an update later that has a specialized variation of this API that optimizes to a memset/memset_pattern. This set of changes is limited to specifically targeting ABI stability and inline characteristics of Data and leaves that optimization as a future addition.

phausler commented:
@swift-ci please test

phausler commented:
@parkera @airspeedswift are we comfortable with the changes here?

parkera commented on stdlib/public/SDK/Foundation/Data.swift:
```diff
@@ -1684,20 +1684,20 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
     
     public subscript(bounds: Range<Index>) -> Data {
-        @inline(__always)
+        @inlinable
```
Just to make sure I understand this part: This is inlinable, but only for Data itself, because others outside the module can't see the backing type?

phausler replied:
hmm it looks like the squash pulled together the revert back to inlines. This has been going back and forth to get appropriate numbers for perf impact so I will have to re-apply the inline removal it looks.

phausler committed:
Remove additional inlines

phausler commented:
@swift-ci please benchmark

parkera commented on stdlib/public/SDK/Foundation/Data.swift:
```diff
@@ -1585,7 +1585,7 @@ public struct Data : ReferenceConvertible, Equatable, Hashable, RandomAccessColl
     }
     
-    @inline(__always)
+    @inlinable
```
Is this inlinable still supposed to be here? (and lines 1604, 1654)

phausler committed:
Remove last public inlines

phausler commented:
@swift-ci please benchmark

phausler commented:
@shahmishal what is happening here? I posted the request to test and merge ~24hrs ago.

phausler commented:
@swift-ci please test and merge

shahmishal commented:
@swift-ci please test and merge
